pinot-core/src/main/java/com/linkedin/pinot/core/query/selection/SelectionOperatorService.java

```diff
@@ -15,369 +15,153 @@
-        .put(DataType.LONG, new DecimalFormat("####################", DecimalFormatSymbols.getInstance(Locale.US)));
-    DEFAULT_FORMAT_STRING_MAP
-        .put(DataType.FLOAT, new DecimalFormat("##########.#####", DecimalFormatSymbols.getInstance(Locale.US)));
-    DEFAULT_FORMAT_STRING_MAP.put(DataType.DOUBLE,
-        new DecimalFormat("####################.##########", DecimalFormatSymbols.getInstance(Locale.US)));
-    DEFAULT_FORMAT_STRING_MAP
-        .put(DataType.INT_ARRAY, new DecimalFormat("##########", DecimalFormatSymbols.getInstance(Locale.US)));
-    DEFAULT_FORMAT_STRING_MAP.put(DataType.LONG_ARRAY,
-        new DecimalFormat("####################", DecimalFormatSymbols.getInstance(Locale.US)));
-    DEFAULT_FORMAT_STRING_MAP
-        .put(DataType.FLOAT_ARRAY, new DecimalFormat("##########.#####", DecimalFormatSymbols.getInstance(Locale.US)));
-    DEFAULT_FORMAT_STRING_MAP.put(DataType.DOUBLE_ARRAY,
-        new DecimalFormat("####################.##########", DecimalFormatSymbols.getInstance(Locale.US)));
-  }
+  private int _numDocsScanned = 0;
 
-  public SelectionOperatorService(Selection selections, IndexSegment indexSegment) {
-    _indexSegment = indexSegment;
-    if ((selections.getSelectionSortSequence() == null) || selections.getSelectionSortSequence().isEmpty()) {
-      _doOrdering = false;
+  /**
+   * Constructor for <code>SelectionOperatorService</code> with {@link IndexSegment}. (Server side)
+   *
+   * @param selection selection query.
+   * @param indexSegment index segment.
+   */
+  public SelectionOperatorService(Selection selection, IndexSegment indexSegment) {
+    _selectionColumns = SelectionOperatorUtils.getSelectionColumns(selection.getSelectionColumns(), indexSegment);
+    _sortSequence = selection.getSelectionSortSequence();
+    _doOrdering = _sortSequence != null && !_sortSequence.isEmpty();
+    int selectionSize = selection.getSize();
+    if (selectionSize == 0) {
+      // No need to select any row.
+      _selectionOffset = 0;
+      _maxRowSize = 0;
     } else {
-      _doOrdering = true;
+      // Select rows from offset to offset + size.
+      _selectionOffset = selection.getOffset();
+      _maxRowSize = _selectionOffset + selectionSize;
     }
-    _sortSequence = selections.getSelectionSortSequence();
-    _selectionColumns = getSelectionColumns(selections.getSelectionColumns());
-    _selectionSize = selections.getSize();
-    _selectionOffset = selections.getOffset();
-    _maxRowSize = _selectionOffset + _selectionSize;
-    _dataSchema = getDataSchema(_sortSequence, _selectionColumns, _indexSegment);
-    _rowComparator = getComparator(_sortSequence, _dataSchema);
+    _dataSchema = SelectionOperatorUtils.extractDataSchema(_sortSequence, _selectionColumns, indexSegment);
     if (_doOrdering) {
-      _rowEventsSet = new PriorityQueue<Serializable[]>(_maxRowSize, _rowComparator);
+      _rowEventsSet = new PriorityQueue<>(Math.max(_maxRowSize, 1), getComparator());
```

Comment:
A better/cleaner way to address the issue with PriorityQueue not allowing constructor with size zero would be to move the check for LIMIT 0 up the stack (ie code should not even reach here). You can choose to address that in a separate PR. If so, please add a comment explaining the issue (why you need to use a queue of size at least 1).